rust/foxglove/src/websocket/service/response.rs

```diff
@@ -0,0 +1,164 @@
+
+use bytes::Bytes;
+use parking_lot::Mutex;
+use tokio_tungstenite::tungstenite::Message;
+use tracing::error;
+
+use crate::websocket::protocol;
+use crate::websocket::service::{CallId, ServiceId};
+
+#[cfg(test)]
+mod tests;
+
+/// The default maximum number of concurrent requests from a particular client.
+const DEFAULT_RESPONSE_CHANNEL_CAPACITY: usize = 32;
+
+/// A bounded response channel.
+pub(crate) struct ResponseChannel {
+    tokens: Arc<Mutex<usize>>,
+    tx: flume::Sender<Response>,
+    rx: flume::Receiver<Response>,
+}
+impl Default for ResponseChannel {
+    fn default() -> Self {
+        Self::new(DEFAULT_RESPONSE_CHANNEL_CAPACITY)
+    }
+}
+impl ResponseChannel {
+    /// Constructs a new service response channel.
+    ///
+    /// The service response channel is configured with a capacity, which represents the maximum
+    /// number of concurrent service calls from a particular client.
+    pub fn new(capacity: usize) -> Self {
+        let tokens = Arc::new(Mutex::new(capacity));
+        let (tx, rx) = flume::bounded(capacity);
+        Self { tokens, tx, rx }
+    }
+
+    /// Removes a response message from the queue.
+    pub async fn next_message(&self) -> Message {
+        // By converting the `Response` to a `Message`, we're also dropping the `Token`.
+        self.rx
+            .recv_async()
+            .await
+            .map(Message::from)
+            .expect("We're holding the sender")
+    }
+
+    /// Removes all responses from the channel.
+    pub fn drain(&self) {
+        self.rx.drain();
+        #[cfg(debug_assertions)]
+        assert_eq!(*self.tokens.lock(), self.tx.capacity().unwrap());
+    }
+
+    /// Allocates a response token and constructs a [`Responder`].
+    ///
+    /// Returns `None` if there are too many concurrent requests.
+    ///
+    /// The responder holds ownership of a [`Token`], which represents a reservation in the
+    /// bounded response channel. This token rides along into the [`Response`], and is finally
+    /// dropped when it is removed from the queue.
+    pub fn prepare_response(
+        &self,
+        service_id: ServiceId,
+        call_id: CallId,
+        encoding: &str,
+    ) -> Option<Responder> {
+        let mut tokens = self.tokens.lock();
+        if *tokens > 0 {
+            *tokens -= 1;
+            Some(Responder {
+                service_id,
+                call_id,
+                encoding: encoding.to_string(),
+                tx: self.tx.clone(),
+                token: Token(self.tokens.clone()),
+            })
+        } else {
+            None
+        }
+    }
+}
+
+/// Represents a reservation for the client's service response queue.
+///
+/// Increments the inner counter (a reference to [`ResponseChannel::tokens`]) when dropped.
+struct Token(Arc<Mutex<usize>>);
+
+impl Drop for Token {
+    fn drop(&mut self) {
+        let mut tokens = self.0.lock();
+        #[cfg(debug_assertions)]
+        assert_ne!(*tokens, usize::MAX);
```

Comment:
debug_assert_ne!

but not required because += 1 is checked_add in a debug build anyway and will panic on overflow